### PR TITLE
Allow disabling LDAP referral chasing on Linux.

### LINKED ISSUE #44826: LDAP referral chasing cannot be disabled on Linux
I have an LDAP server running on a separate machine. 

I can run the following code successfully when using the dotnet 5.0.100 SDK on Windows 10, and I can perform the corresponding query successfully using `ldapsearch` on Ubuntu 20.04.1 LTS. However, running the following code on Ubuntu 20.04.1 LTS or in a Docker container built with the dotnet 5.0 SDK image that also has `libldap-common` installed fails as described below. 

I am attempting to use System.DirectoryServices.Protocol to perform an LDAP query that I was able to perform successfully previously when I was using the Novell library. Specifically, I am creating the connection and binding using
```
_ldapConnection = new LdapConnection(
                new LdapDirectoryIdentifier(host, port, false, false),
                new NetworkCredential(userName: "CN=Administrator,CN=Users,DC=inc,DC=company,DC=com", password: "mypassword"),
                AuthType.Basic
            );
_ldapConnection.Bind();
```
And then doing the following query

```
string searchBase = "DC=inc,DC=company,DC=com";
            string filter = "(&(objectClass=person)(objectClass=user))";
            var search = new SearchRequest(searchBase, filter, SearchScope.Subtree, null);
            var response = (System.DirectoryServices.Protocols.SearchResponse)_ldapConnection.SendRequest(search, System.TimeSpan.FromMinutes(1));
```
However, doing so results in the following error
```
Unhandled exception. System.DirectoryServices.Protocols.DirectoryOperationException: An unspecified operation error occurred.
   at System.DirectoryServices.Protocols.LdapConnection.ConstructResponseAsync(Int32 messageId, LdapOperation operation, ResultAll resultType, TimeSpan requestTimeOut, Boolean exceptionOnTimeOut, Boolean sync)
   at System.DirectoryServices.Protocols.LdapConnection.SendRequest(DirectoryRequest request, TimeSpan requestTimeout)
```
I unpacked every property in the exception itself, and in the corresponding response attached to the exception. This is what I found:
```
Data: System.Collections.ListDictionaryInternal
Help Link: 
HResult: -2146233088
InnerException: 
Message: An unspecified operation error occurred.
Source: System.DirectoryServices.Protocols
StackTrace:    at System.DirectoryServices.Protocols.LdapConnection.ConstructResponseAsync(Int32 messageId, LdapOperation operation, ResultAll resultType, TimeSpan requestTimeOut, Boolean exceptionOnTimeOut, Boolean sync)
   at System.DirectoryServices.Protocols.LdapConnection.SendRequest(DirectoryRequest request, TimeSpan requestTimeout)
   at Keep_LdapCore.LdapService.LogUsers() in /workspaces/Keep-LdapCore/LdapService.cs:line 35
TargetSite: Void MoveNext()
Response.Controls: System.DirectoryServices.Protocols.DirectoryControl[]
Response.Error Message: Referral:
ldap://ForestDnsZones.inc.company.com/DC=ForestDnsZones,DC=inc,DC=company,DC=com
ldap://DomainDnsZones.inc.company.com/DC=DomainDnsZones,DC=inc,DC=company,DC=com
ldap://inc.company.com/CN=Configuration,DC=inc,DC=company,DC=com
Response.MatchedDN: 
Response.Referral: System.Uri[]
Response.RequestId: 
Response.Result Code: -1
```

### PR DESCRIPTION
**Pull request by iinuwa**
Before this, changing SessionOptions.ReferralChasing on Linux was ineffective, since we were passing `ref int` to OpenLDAP, which does not follow the pointer passed and interprets any non-zero value as "enabled".

This passes a boolean directly instead, which the library is able to detect properly.

This is a little bit of a hack, since it relies on the fact that [OpenLDAP doesn't actually check for the correct value of `LDAP_OPT_ON`](https://git.openldap.org/openldap/openldap/-/blob/OPENLDAP_REL_ENG_2_4_58/libraries/libldap/options.c#L470-477), which changes every time the program is run. Instead, it compares whether the passed value is 0 when cast to `int *`, which makes it `false`; passing any other value evaluates to `true`. Actually getting the correct value of LDAP_OPT_ON, I think would require a putting a get_ldap_opt_on() function in a native shim for OpenLDAP, which seems a little overkill. The [documentation](https://www.openldap.org/software/man.cgi?query=ldap_get_option&sektion=3&apropos=0&manpath=OpenLDAP+2.4-Release) says

>  its value should either be `LDAP_OPT_OFF` or `LDAP_OPT_ON`

but maybe it's ok to fudge that.

Fixes #44826.

### CODE REVIEW

File: src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs
Comment (danmoseley) [inline src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs:36]: Please add to resources.

File: src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs
Comment (danmoseley) [inline src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs:34]: Nit: I personally find `a != x && b != y` easier to read than `!(a == x || b == y)`

**Comment by iinuwa** [comment]
@danmoseley or @krwq, do you have any more feedback for this PR?

**Comment by danmoseley** [comment]
I do not. @buyaa-n @joperezr @krwq own this area and @joperezr is out. Can one of you please review this and merge as appropriate?

BTW these three folks are @dotnet/area-system-directoryservices

**Review by jeffhandley — APPROVED** [review]

**Comment by jeffhandley** [comment]
@buyaa-n is also out right now; it looks good to me though.

The check failures look unrelated, but I'm going to rerun them once before merging.

**Comment by iinuwa** [comment]
One thing to note: this PR does not enable controlling the referral types followed (e.g. subordinate or external); it's all or nothing. I believe it is possible to configure the type of referrals followed by setting the [`LDAP_CONTROL_REFERRALS`](https://git.openldap.org/openldap/openldap/-/blob/OPENLDAP_REL_ENG_2_4_58/doc/drafts/draft-ietf-ldapext-ldap-c-api-xx.txt#L1307-1347) client control globally using the [`LDAP_OPT_CLIENT_CONTROLS`](https://git.openldap.org/openldap/openldap/-/blob/OPENLDAP_REL_ENG_2_4_58/doc/drafts/draft-ietf-ldapext-ldap-c-api-xx.txt#L1098-1105), which is a [NULL-terminated arrray](https://git.openldap.org/openldap/openldap/-/blob/OPENLDAP_REL_ENG_2_4_58/doc/drafts/draft-ietf-ldapext-ldap-c-api-xx.txt#L1295-1296) of `LDAPControl`s option on the LDAP handle.

I think that this can be handled in a separate PR though.

**Review by krwq — APPROVED** [review]
This conceptually looks good but I'm not really familiar with this space... I'm ok with merging this

File: src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs
Comment (krwq) [inline src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs:4]: Are all the new napsaces needed here and in the other .cs file?

Comment (iinuwa) [reply src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs:4]: Oh, right. `System.ComponentModel` is needed for `InvalidEnumArgumentException` in both files. The other two in `LdapSessionOptions.Linux.cs` are not needed anymore. I'll remove them.

**Comment by iinuwa** [comment]
@dotnet/area-system-directory-services: Thanks for looking this over! Do PRs in this area need to be approved by all three of you, or does the auto-merge label need to be set?

File: src/libraries/System.DirectoryServices.Protocols/tests/LdapSessionOptionsTests.cs
Comment (krwq) [inline src/libraries/System.DirectoryServices.Protocols/tests/LdapSessionOptionsTests.cs:17]: this test seems identical to the one below, can you join them together? `[PlatformSpecific(TestPlatforms.Windows | TestPlatforms.Linux)]` should work

**Comment by krwq** [comment]
@iinuwa this change looks correct to me but I'm not too familiar with LDAP, @joperezr is but he's currently on vacation (until around August). We can go about it two ways:
- I merge this now (after the last comment) and ask @joperezr to take a second look after he's back from vacation in case he spots something we get it fixed later
- we wait with this PR for @joperezr to be back

considering this change is scoped to Linux and the referral chasing doesn't seem to work there I personally think it's acceptable to merge this. One thing to note is that we're really close to ask mode (branch locking before release - after that all changes need to be approved before merging to mitigate risk of breaking something) and that is making me slightly uncomfortable. @jeffhandley are you ok with me merging this now (after last comment related to tests) and @joperezr taking a second look later?

File: src/libraries/System.DirectoryServices.Protocols/tests/LdapSessionOptionsTests.cs
Comment (iinuwa) [reply src/libraries/System.DirectoryServices.Protocols/tests/LdapSessionOptionsTests.cs:17]: They have the same content, but different inputs, since setting `External` is not supported on Linux but it is on Windows. Can we scope the `InlineData` attribute for each platform? (Sorry if that's obvious: I didn't see examples of this elsewhere.)

**Pull request merged**
